### PR TITLE
feat(core): special case the github npm registry when migrating

### LINKED ISSUE #10223: github npm registry does not allow custom fields on `npm view`, breaking migrations
## Current Behavior

[Discussion on slack](https://nrwlcommunity.slack.com/archives/CMFKWPU6Q/p1650630724184429).

`nx migrate` relies on `npm view` to get information about migrations to run from updated packages, specifically by getting the `ng-update` or `nx-migration` fields.

However, in the case of packages that reside in the github package registry, its implementation of `npm view` does not return custom fields.

In that case, [getting the package with npm view](https://github.com/nrwl/nx/blob/master/packages/nx/src/command-line/migrate.ts#L522) will not fail since `npm view` does not throw, so it does not fall back to [getting the package with package-manager install](https://github.com/nrwl/nx/blob/master/packages/nx/src/command-line/migrate.ts#L527).

This results in migrations not being picked up.

## Expected Behavior

All packages, including those under the github package registry, should allow migrations.

I've been in contact with the github support team, and it's been communicated that exposing arbitrary package.json fields is not prioritized and will not be implemented any time soon.

Given that, my suggestion is to special-case the npm github package registry to opt out of the `npm view` behavior.

### Failure Logs

There are no failures, as `npm view`, in this case, acts as if the package has no migrations.

### Environment

This is not an NX issue, so it will happen on all versions with the `npm view` behavior.

### PR DESCRIPTION
**Pull request by gioragutt**
Fixes #10223

### CODE REVIEW

File: packages/nx/src/command-line/migrate.ts [hunk 1]
```diff
@@ -487,72 +488,99 @@ function versions(root: string, from: Record<string, string>) {
 
 // testing-fetch-start
 function createFetcher() {
-  const migrationsCache: Record<
-    string,
-    Promise<ResolvedMigrationConfiguration>
-  > = {};
-  const resolvedVersionCache: Record<string, Promise<string>> = {};
+  const migrationsCache: Record<string, ResolvedMigrationConfiguration> = {};
+  const resolvedVersionCache: Record<string, string> = {};
```
Comment (AgentEnder) [inline packages/nx/src/command-line/migrate.ts:492]: Is there a reason for changing these away from the Promise<T> signatures? IIRC there was a reason behind it.

File: packages/nx/src/command-line/migrate.ts [hunk 2]
```diff
@@ -487,72 +488,99 @@ function versions(root: string, from: Record<string, string>) {
...
+
+    if (!packageRegistryCache[packageName]) {
+      const { stdout } = await execAsync(
+        `${pm} config get ${configKeyPrefix}registry`
+      );
+
+      packageRegistryCache[configKeyPrefix] = stdout.trim();
+    }
+    return packageRegistryCache[configKeyPrefix];
   }
```
Comment (AgentEnder) [inline packages/nx/src/command-line/migrate.ts:523]: Lets move this outside of the function scope, since its not dependent on any scoped vars. It **_may_** be worth caching this too, not sure.

File: packages/nx/src/command-line/migrate.ts [hunk 3]
```diff
@@ -487,72 +488,99 @@ function versions(root: string, from: Record<string, string>) {
...
-      return migrationsCache[`${packageName}-${packageVersion}`];
+    if (
+      (await getRegistryOfPackage(packageName)) ===
+      'https://npm.pkg.github.com/'
+    ) {
+      /**
+       * The github npm package registry does not yet support custom fields,
+       * meaning that `npm view` will not return the nx migration information,
+       * so we have to fallback to getting the information with a full install.
+       */
```
Comment (AgentEnder) [inline packages/nx/src/command-line/migrate.ts:537]: nit: Move this above the if statement, s.t. the condition is closer to the logic

File: packages/nx/src/command-line/migrate.ts [hunk 4]
```diff
@@ -487,72 +488,99 @@ function versions(root: string, from: Record<string, string>) {
...
           delete result.schematics;
         }
-        resolvedVersion = result.version;
         return result;
       }
     );
 
-    setCache(packageName, packageVersion);
+    setMigrationCache(packageName, packageVersion, migrations);
+    setMigrationCache(packageName, migrations.version, migrations);
```
Comment (AgentEnder) [inline packages/nx/src/command-line/migrate.ts:583]: What's the reason behind setting both keys here?

**Review by AgentEnder — CHANGES_REQUESTED** [review]
Hey! A few small questions + nits, overall it doesn't seem like anything is overtly wrong just want to cover some bases.

File: packages/nx/src/command-line/migrate.ts (hunk 1 above)
Comment (gioragutt) [reply packages/nx/src/command-line/migrate.ts:492]: Overall my problem with the current implementation is that the flow is very hard to follow. The `setCache` method relies on some side-effect since it doesn't pass what to cache.

I'd revert the whole thing back if needed, but overall it was very hard for me to follow.
I kinda understand now what it does, but the flow was hard to grasp.

I assume that it was to make sure that the result is cached ASAP, without getting the migration itself.
I'll try to implement this without this refactor.

File: packages/nx/src/command-line/migrate.ts (hunk 4 above)
Comment (gioragutt) [reply packages/nx/src/command-line/migrate.ts:583]: This is to replicate the behavior from before. It'd cache the version specified and also the resolved version.

File: packages/nx/src/command-line/migrate.ts (hunk 1 above)
Comment (AgentEnder) [reply packages/nx/src/command-line/migrate.ts:492]: I agree that the current flow is a bit hard to follow, but I think we ran into numerous race conditions when debugging w/o using promises in the cache. 

I'd recommend falling back on the prev impl for these

**Comment by amirtoole** [comment]
Hey; any updates on this PR? We're running into https://github.com/nrwl/nx/issues/10223 as well.
Thanks

**Comment by AgentEnder** [comment]
@gioragutt I'm going to close this out. Thanks for the effort than went in to this, and I appreciate that it wasn't super trivial, but I don't think there is a high amount of value in adding a special case like this.

We may look into it later and see if there is a more minimal approach that could fix this.
